text/3803-enum-repr-no-niches.md

```diff
@@ -0,0 +1,709 @@
+        // and can thus represent all bit-patterns of `u8`.
+        unsafe { core::mem::transmute::<u8, Weather>(value) }
+    }
+}
+
+assert_eq!(Weather::from(0), Weather::Sunny);
+assert_eq!(Weather::from(1), Weather::Windy);
+assert_eq!(Weather::from(2), Weather::Rainy);
+
+let snowy = Weather::from(3);
+assert_ne!(snowy, Weather::Sunny);
+assert_ne!(snowy, Weather::Windy);
+assert_ne!(snowy, Weather::Rainy);
+
+assert_eq!(size_of::<Weather>(), size_of::<u8>());
+// No niches for the `None` to fit in.
+assert_ne!(size_of::<Weather>(), size_of::<Option<Weather>>());
+```
+
+And thus, using the enum at the FFI boundary is similarly possible.
+
+## `as`-casting
+
+Casting from the enum to the underlying integer continues to access the discriminant [as described in the reference](https://doc.rust-lang.org/1.86.0/reference/items/enumerations.html#casting). Building on the previous example:
+
+```rust
+assert_eq!(Weather::Sunny as u8, 0);
+assert_eq!(snowy as u8, 3);
+```
+
+Note that this RFC does not propose a way to safely construct enums of the unknown variants, see [the future possibilities section][future-possibilities] for that.
+
+## Discriminant
+
+The `core::mem::Discriminant` of the enum exposed to the user should differ based on the actual underlying value, regardless of the variants not being available in the source code. Building on the previous example:
+
+```rust
+use core::mem::discriminant;
+assert_eq!(discriminant(&snowy), discriminant(&snowy));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::Sunny));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::Windy));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::Rainy));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::from(42)));
+```
+
+## Usage in pattern matching
+
+Adding the `no_niches` modifier affects exhaustiveness-checking, since it adds extra variants.
+
+Unlike `#[non_exhaustive]`, the check is enabled everywhere, regardless of visibility attributes and crate boundaries (because it works on the ABI level).
+
+Whether exhaustiveness-checking is affected by `no_niches` enums that already has enough variants to represent every bit-pattern of the underlying integer type depends on the final syntax, see [the unresolved questions section][unresolved-questions].
+
+## Derives
+
+`#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]` compares the discriminant, regardless of the source-ordering (like they do today).
+
+`#[derive(Default)]` continues to require that a variant is marked with `#[default]`.
+
+The exact output of `#[derive(Debug)]` is [an unresolved question][derivedebug].
+
+User-defined derives and proc-macros will have to know when to emit the extra match arm (so most user-defined derive macros probably won't support `#[repr(_, no_niches)]` initially). Proc-macros could either choose to inspect the `repr`, or to defensively emit an extra `#[allow(unreachable_patterns)] _ => {}` match arm.
```

Comment:
If it doesn't exist already, it would be useful to publish a zero-dependency crate to inspect reprs and match at least the most common ones so that not every proc macro author needs to reinvent the wheel. Such a crate could then be mentioned here.

Reply:
Interesting idea!

I've added it to the future possibilities section, since I do not intend to maintain such a crate myself (and since I suspect it'd be a better idea for macro authors to defensively emit the extra match arm).